### ISSUE TITLE
Blocked CNAME of allowed domain not able to easily be allowed

### ISSUE THREAD

**Issue by CJ-Ross** [opnsense/core#6722]
**Important notices**

Before you add a new report, we ask you kindly to acknowledge the following:

- [x ] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md
- [ x] I am convinced that my issue is new after having checked both open and closed issues at https://github.com/opnsense/core/issues?q=is%3Aissue

**Describe the bug**

I've tested this with 23.1.11 but  I don't believe any changes in 23.7 would have affected this.  If the CNAME of an allowed domain is on a blocklist, there is not a simple way to allow that query to successfully complete.

**To Reproduce**

1. Enable DNSBL in Unbound using the Steven Black list.
2. Enable Unbound Reporting
3. Query click.redditmail.com
4. Visit Unbound Reporting and whitelist click.redditmail.com
5. Configure Unbound to flush cache on restart and restart Unbound
6. Query click.redditmail.com
7. Visit Unbound Reporting and see that A and AAAA records are allowed but CNAME is blocked
8. CNAME shows as already whitelisted but query for click.reddit.com still fails

**Expected behavior**

Whitelisted click.redditmail.com should resolve

**Describe alternatives you considered**

Current workaround is to manually query the allowed domain, get the resulting CNAME and then whitelist that in the DNSBL screen.  Alternative would be for Unbound Reporting to show the both the original and resulting CNAME and/or recognize that the CNAME is what is still blocked and therefore show the whitelist option instead of block.

Probably the least feasible option would be for Unbound to realize that the original query is allowed and therefore not check the resulting CNAME against the blocklist.  This would likely cause more problems than it solves as well as reducing security provided by the blocklist.

**Screenshots**

https://imgur.com/N3dFp03
https://imgur.com/79oddsl

**Additional context**

See thread for more context.  https://forum.opnsense.org/index.php?topic=35218.0

**Comment by swhite2** [comment]
>Probably the least feasible option would be for Unbound to realize that the original query is allowed and therefore not check the resulting CNAME against the blocklist. This would likely cause more problems than it solves as well as reducing security provided by the blocklist.

This is entirely correct.

> Alternative would be for Unbound Reporting to show the both the original and resulting CNAME and/or recognize that the CNAME is what is still blocked and therefore show the whitelist option instead of block.

The statistics are from the perspective of the incoming queries, therefore showing two entries here will likely pollute the counters. The current code is somewhat aware of this situation, but pretends to be the original query so domains don't appear out of nowhere for the person reviewing the logs. The easiest solution would likely be to drop logging the incoming query (e.g. `click.redditmail.com`) and instead log `<resulting CNAME> (CNAME of <original queried domain>)`.

**Comment by CJ-Ross** [comment]
> > Alternative would be for Unbound Reporting to show the both the original and resulting CNAME and/or recognize that the CNAME is what is still blocked and therefore show the whitelist option instead of block.
> 
> The statistics are from the perspective of the incoming queries, therefore showing two entries here will likely pollute the counters. The current code is somewhat aware of this situation, but pretends to be the original query so domains don't appear out of nowhere for the person reviewing the logs. The easiest solution would likely be to drop logging the incoming query (e.g. `click.redditmail.com`) and instead log `<resulting CNAME> (CNAME of <original queried domain>)`.

Isn't that exactly what I listed as the first alternative?

**Comment by swhite2** [comment]
I am simply eliminating ambiguity.. "exactly" does not quite fit.

**Comment by CJ-Ross** [comment]
I've discovered another issue related to Unbound DNSBL and CNAME records.  I'm not sure if it needs to be opened as a separate issue or rolled into this one.

https://forum.opnsense.org/index.php?topic=36068.0

The TLDR is that CNAME records that resolve to a CNAME record appear to be broken.

**Comment by AJSNetworks** [comment]
I am also experiencing the same issue. Some domains in my "Whitelist" work fine but the most recent one added does not, regardless of restarts and waiting out TTL.

Interesting observations that I hope can shed light on this:
- REPORTING: UNBOUND DNS logs it as a "block", but the "Command" button only offers to "Block Domain", when "Whitelist Domain" is expected
- Resolution issues coming from the browser with the following error "DNS_PROBE_POSSIBLE"
- Using nslookup (windows) does resolve the domain in question (I see it as a block in REPORTING: UNBOUND DNS
- Domain in question: "learn.cisecurity.org", being blocked from at least AdAway and EasyPrivacy lists

**Comment by oliveiraleo** [comment]
Hello, I'm currently using the versions listed below and I can confirm this bug

```
OPNsense 23.7.11-amd64
FreeBSD 13.2-RELEASE-p7
OpenSSL 1.1.1w
```

The domain I tried to whitelist is `mparticle.weather.com` but unbound still blocks the requests with the same behavior reported by @CJ-Ross and @AJSNetworks even after whitelisting and reloading the service

**Comment by gforgas** [comment]
Hello, 

Running same version reported by OP here and same issue. 

`OPNsense 23.1.11_1-amd64
FreeBSD 13.1-RELEASE-p8
OpenSSL 1.1.1u 30 May 2023` 

In this case the domain that is not being whitelisted is `go.missionbio.com.	2250	IN	CNAME	go.pardot.com.`

Disabling DNSBL "solves" the issue, but is not really what we want.

**Comment by CJ-Ross** [comment]
I have plans to work on this issue along with some other improvements to the whole DNSBL handling and reporting but with everything else I have going on I have no idea when I'll be able to dedicate time to it.

In the meantime, anyone coming across this issue can use a DNS lookup tool(local or online) to find out what the next CNAME in the chain is until they get to the actual domain and unblock each one of those.  While cumbersome, it does work to allow you to access something that is on a downloaded DNSBL.

**Comment by nemxwasp** [comment]
I must be doing something wrong. In order to watch live sports on Paramount+ I have to allow pubads.g.doubleclick.net. I added it and its other cname pubads46.g.doubleclick.net to whitelist domains and its still blocked by DNSBL. I don't watch often but when I do I have to disable blocklist altogether and clear my local dns cache to get it to work.

**Comment by xm4rcell0x** [comment]
The problem is still here...
Any news?

**Comment by nraudigy2bf** [comment]
Same issue in OPNSense version 25.1

**Comment by CompPhy** [comment]
I'm on 25.1 and this seems even worse than the original description.  The notes here seem to indicate that whitelisting both the CNAME, and A name, should allow things to work.  However, that does not appear to be the case for me.  I'm having to actually disable blocklists at this point because every query continues to indicate "Block" even after restarting, flushing cache, etc. multiple times.

Here's my whitelist, try it for yourself:

![Image](https://github.com/user-attachments/assets/97e21aed-e02b-447c-9513-18a431cb4411)

**Comment by nicholastripp** [comment]
I'm also still experience this issue on OPNsense 25.1.7_2. The workaround listed here does work, but it'd be great if it worked properly. I'm sure there are others experiencing this problem that won't find this thread.

**Comment by DonutPirate** [comment]
At first glance this issue still appears to be present in 25.7. Was only having issues with CNAMEs specifically not honoring the Unbound whitelist. Frustrating to troubleshoot for someone new to OPNsense; I assumed I had misconfigured something myself.

**Comment by Baenwort** [comment]
I'm trying to unblock the cdn.fastly.steamstatic.com which goes to shared.valve.map.fastly.net but even after adding both to the whitelist I am unable to load Steam Avatars.

This whitelist not over ruling the block list defeats the entire purpose of having an allow list. Why is there no way to over rule the blocklist?

**Comment by nemxwasp** [comment]
Did you try the regex from my post above? I don't know regex at all but this always seems to unblock it for me provided I add all the different cnames.

(.*)?(\.)?shared.valve.map.fastly.net
(.*)?(\.)?cdn.fastly.steamstatic.com

**Comment by swhite2** [comment]
While revisiting this area, I've pushed https://github.com/opnsense/core/commit/854200a1125881609bf349af6217455f35757b3d. If anyone wants to test this:

```
# opnsense-patch 854200a
```

(restart Unbound afterwards)

This should automatically allow any CNAME associated with a given whitelist entry.
